source/common/protobuf/message_validator_impl.h

```diff
@@ -14,32 +14,42 @@ class NullValidationVisitorImpl : public ValidationVisitor {
 public:
   // Envoy::ProtobufMessage::ValidationVisitor
   void onUnknownField(absl::string_view) override {}
+  void onDeprecatedField(absl::string_view) override {}
 };
 
 ValidationVisitor& getNullValidationVisitor();
 
 class WarningValidationVisitorImpl : public ValidationVisitor,
                                      public Logger::Loggable<Logger::Id::config> {
 public:
-  void setCounter(Stats::Counter& counter);
+  void setUnknownCounter(Stats::Counter& counter);
+  void setDeprecatedCounter(Stats::Counter& counter);
 
   // Envoy::ProtobufMessage::ValidationVisitor
   void onUnknownField(absl::string_view description) override;
+  void onDeprecatedField(absl::string_view description) override;
 
 private:
+  void onUnexpectedField(absl::string_view description, Stats::Counter* counter,
+                         const ValidationType& validation_type);
   // Track hashes of descriptions we've seen, to avoid log spam. A hash is used here to avoid
   // wasting memory with unused strings.
   absl::flat_hash_set<uint64_t> descriptions_;
-  // This can be late initialized via setCounter(), enabling the server bootstrap loading which
-  // occurs prior to the initialization of the stats subsystem.
-  Stats::Counter* counter_{};
-  uint64_t prestats_count_{};
+  // This can be late initialized via setUnknownCounter(), enabling the server bootstrap loading
+  // which occurs prior to the initialization of the stats subsystem.
+  Stats::Counter* unknown_counter_{};
+  // This can be late initialized via setDeprecatedCounter(), enabling the server bootstrap loading
+  // which occurs prior to the initialization of the stats subsystem.
+  Stats::Counter* deprecated_counter_{};
+  uint64_t prestats_unknown_count_{};
+  uint64_t prestats_deprecated_count_{};
```

Comment:
adding extra counter for deprecated fields may not be needed, as there is stats counter in Runtime already: `deprecated_feature_use`

Reply:
Yep, that's fair enough.

Reply:
True, it's a bit strange that we bump the stat in runtime rather than in the calling code, but I don't think we can undo that at this point as folks are likely relying on it. So, yeah, we can simplify the code here quite a bit as a result.